File: leetcode/src/ConvertSortedArrayToBinarySearchTree.java
public class ConvertSortedArrayToBinarySearchTree extends BaseTreeNode {
    public static void main(String[] args) {
        ConvertSortedArrayToBinarySearchTree instance = new ConvertSortedArrayToBinarySearchTree();
        printTreeNode(instance.sortedArrayToBST(new int[]{-10,-3,0,5,9}));
        printTreeNode(instance.sortedArrayToBST(new int[]{1,3}));
        printTreeNode(instance.sortedArrayToBST(new int[]{-1,0,1,2}));
        printTreeNode(instance.sortedArrayToBST(new int[]{0,1,2,3,4,5,6,7}));
    }
    public TreeNode sortedArrayToBST(int[] nums) {
        return generateTreeNode(nums, 0, nums.length - 1);
    }
    private TreeNode generateTreeNode(int[] nums, int start, int end) {
        TreeNode result = null;
        if (start == end) {
            result = new TreeNode(nums[start]);
        } else if (start + 1 == end) {
            result = new TreeNode(nums[start], null, new TreeNode(nums[end]));
        } else if (start + 2 == end) {
            result = new TreeNode(nums[start + 1], new TreeNode(nums[start]), new TreeNode(nums[end]));
        } else {
            int middle = (end + start) / 2;
            result = new TreeNode(nums[middle], generateTreeNode(nums, start, middle - 1), generateTreeNode(nums, middle + 1, end));
        }
        return result;
    }
}

File: leetcode/src/python/convert_sorted_array_to_binary_search_tree.py
from typing import Optional
from typing import List

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class ConvertSortedArrayToBinarySearchTree:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        result = None
        l = len(nums)
        if l > 0:
            middle = int((l - 1) / 2)
            result = TreeNode(nums[middle], self.sortedArrayToBST([] if middle == 0 else nums[0:middle]), self.sortedArrayToBST([] if middle == l - 1 else nums[middle + 1:l]))
        return result